### PR TITLE
[Unity] Add `has_function` to RelaxVM

### PR DESCRIPTION
**Pull request by jinhongyii**
add a "has_function" function in executable

### CODE REVIEW

**Comment by jinhongyii** [comment]
cc: @junrushao @tqchen

File: include/tvm/runtime/disco/session.h [hunk 1]
```diff
@@ -195,6 +195,18 @@ class SessionObj : public Object {
    * \param host_array The array to be copied to worker-0
    * \param remote_array The NDArray on worker-0
    */
+
+  DRef CopyToWorker0Wrapper(const NDArray& host_array) {
```
Comment (tqchen) [inline include/tvm/runtime/disco/session.h:199]: To minimize the interface, let us consider only expose remote.Empty.

Then move the wrapper to the upper level applications

File: src/runtime/relax_vm/executable.cc [hunk 2]
```diff
@@ -75,6 +75,10 @@ PackedFunc Executable::GetFunction(const String& name, const ObjectPtr<Object>&
       vm->LoadExecutable(GetObjectPtr<Executable>(this));
       *rv = Module(vm);
     });
+  } else if (name == "func_exists") {
```
Comment (tqchen) [inline src/runtime/relax_vm/executable.cc:78]: LGTM

File: include/tvm/runtime/disco/session.h [hunk 3]
```diff
@@ -236,7 +248,7 @@ class Session : public ObjectRef {
  public:
   /*! \brief Create a session backed by a thread pool of workers */
   static Session ThreadedSession(int num_workers);
-  TVM_DEFINE_MUTABLE_NOTNULLABLE_OBJECT_REF_METHODS(Session, ObjectRef, SessionObj);
+  TVM_DEFINE_MUTABLE_OBJECT_REF_METHODS(Session, ObjectRef, SessionObj);
```
Comment (junrushao) [inline include/tvm/runtime/disco/session.h:251]: Why this change?

File: include/tvm/runtime/disco/session.h (hunk 1 above)
Comment (junrushao) [reply include/tvm/runtime/disco/session.h:199]: `Empty` is exposed on python: https://github.com/apache/tvm/blob/unity/python/tvm/runtime/disco/session.py#L117-L142.

Shall we migrate some of the python pieces to C++?

File: include/tvm/runtime/disco/session.h [hunk 4]
```diff
@@ -155,7 +155,7 @@ class DRefObj : public Object {
  */
 class DRef : public ObjectRef {
  public:
-  TVM_DEFINE_MUTABLE_NOTNULLABLE_OBJECT_REF_METHODS(DRef, ObjectRef, DRefObj);
+  TVM_DEFINE_MUTABLE_OBJECT_REF_METHODS(DRef, ObjectRef, DRefObj);
```
Comment (junrushao) [inline include/tvm/runtime/disco/session.h:158]: If there isn't a particular reason, please keep objects as "not-nullable" as possible

File: include/tvm/runtime/disco/session.h (hunk 3 above)
Comment (jinhongyii) [reply include/tvm/runtime/disco/session.h:251]: In MLC we want to make Session a field. To initialize it, we are using ThreadSession(worker_num) instead of its constructor to create a session, so it has to be nullable to compile. Another reason is that in definition of DRef, we are already having  
``` /*! \brief Back-pointer to the host controler session */
  ObjectRef session{nullptr};
```
So there's no reason session can't be nullable.

Comment (junrushao) [reply include/tvm/runtime/disco/session.h:251]: To be clear, it's not-nullable since it's properly created. If you wanted to cover the case where Session could be a field of another TVM Object, the common paradigm we are using across TVM codebase is to mark it as "initially null":

```C++
Session sess{nullptr};
```

Note that this pattern is widely used in MetaSchedule already.

There are multiple folds of reasons for using not-nullable, as TVM object system will enforce more checking, and also developers of other modules could feel very safe to use an object without worrying it's null.

Comment (jinhongyii) [reply include/tvm/runtime/disco/session.h:251]: sounds good

File: include/tvm/runtime/disco/session.h (hunk 1 above)
Comment (jinhongyii) [reply include/tvm/runtime/disco/session.h:199]: If we want to expose in c++, that's only 2 lines of code. So I don't think it's necessary.

**Comment by jinhongyii** [comment]
please review @junrushao @tqchen

**Review by tqchen — APPROVED** [review]

File: src/runtime/relax_vm/executable.cc (hunk 2 above)
Comment (tqchen) [inline src/runtime/relax_vm/executable.cc:78]: thinking a bit more, how about rename to has_function

**Review by tqchen — COMMENTED** [review]
one minor nit

File: src/runtime/relax_vm/executable.cc (hunk 2 above)
Comment (junrushao) [reply src/runtime/relax_vm/executable.cc:78]: sounds good. I'm updating the name on my end to accelerate PR merge
